aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -39,6 +50,148 @@ fn is_string_type(context: &SafeNativeContext, type_arg: &Type) -> SafeNativeRes
+    let (id, value, limit) = get_aggregator_fields(&ty_args[0], args)?;
+    let aggregator = aggregator_data.get_aggregator(id, limit)?;
+    Ok(smallvec![Value::bool(aggregator.try_add(value).is_ok())])
+}
+
+/***************************************************************************************************
+ * native fun try_sub<Element>(aggregator: &mut Aggregator<Element>, value: Element): bool;
+ **************************************************************************************************/
+fn native_try_sub(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 2);
+
+    context.charge(AGGREGATOR_V2_TRY_SUB_BASE)?;
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+    let (id, value, limit) = get_aggregator_fields(&ty_args[0], args)?;
+    let aggregator = aggregator_data.get_aggregator(id, limit)?;
+    Ok(smallvec![Value::bool(aggregator.try_sub(value).is_ok())])
+}
+
+/***************************************************************************************************
+ * native fun read<Element>(aggregator: &Aggregator<Element>): Element;
+ **************************************************************************************************/
+
+fn native_read(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_READ_BASE)?;
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+
+    match ty_args[0] {
+        Type::U128 => {
+            // Extract information from aggregator struct reference.
+            let (id, limit) = aggregator_value_as_u128(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit)?;
+            let value = aggregator.read_and_materialize(aggregator_context.resolver, &id)?;
+            Ok(smallvec![Value::u128(value)])
+        },
+        Type::U64 => {
+            let (id, limit) = aggregator_value_as_u64(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit as u128)?;
+            let value = aggregator.read_and_materialize(aggregator_context.resolver, &id)?;
+            if value > u64::MAX as u128 {
+                return Err(SafeNativeError::Abort {
+                    abort_code: EREAD_OUT_OF_BOUNDS,
+                });
+            }
+            Ok(smallvec![Value::u64(value as u64)])
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: EUNSUPPORTED_AGGREGATOR_TYPE,
+        }),
+    }
```

Comment:
```
    // Extract information from aggregator struct reference.
    let (id, limit) = pop_aggregator_fields_of_type(args, ty_args[0])?;
    let aggregator = aggregator_data.get_aggregator(id, limit)?;
    let value = aggregator.read_and_materialize(aggregator_context.resolver, &id)?;
           
    if value > limit {
        return Err(SafeNativeError::Abort {
             abort_code: EREAD_OUT_OF_BOUNDS,
        });
    }

    match ty_args[0] {
        Type::U128 => {
            Ok(smallvec![Value::u128(value)])
        },
        Type::U64 => {
            // this should never happen, but since we check the limit above - this should really never happen
            if value > u64::MAX as u128 {
                return Err(SafeNativeError::Abort {
                    abort_code: EREAD_OUT_OF_BOUNDS,
                });
            }
            Ok(smallvec![Value::u64(value as u64)])
        },
        _ => Err(SafeNativeError::Abort {
            abort_code: EUNSUPPORTED_AGGREGATOR_TYPE,
        }),
    }
```

Reply:
Done.